06-TimeManagement/my_time.c

```diff
@@ -0,0 +1,170 @@
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/device.h>
+#include <linux/err.h>
+#include <linux/ctype.h>
+#include <linux/string.h>
+#include <linux/jiffies.h>
+#include <linux/time.h>
+#include <linux/timer.h>
+
+#define DEFAULT_DELAY_IN_MS 5000
+#define MIN_DELAY_IN_MS 100
+
+static unsigned long jiffies_stamp;
+static unsigned long jiffies_now;
+static unsigned long delta_in_ms;
+static struct timespec epoch_now;
+static struct timer_list my_timer;
+
+
+static void mod_my_timer(void)
+{
+	mod_timer(&my_timer, jiffies + msecs_to_jiffies(delta_in_ms));
+}
+
+static void my_timer_callback(unsigned long data)
+{
+	pr_info("my_time: call %s, delay: %lu [ms]\n", __func__, delta_in_ms);
+	mod_my_timer();
+}
+
+static ssize_t my_show(struct class *class, struct class_attribute *attr,
+						char *buf)
+{
+	struct tm last_read;
+
+	pr_info("my_time: call %s", __func__);
+
+	jiffies_now = jiffies;
+
+	if (time_before(jiffies_stamp, jiffies_now)) {
+		getnstimeofday(&epoch_now);
+		time_to_tm(epoch_now.tv_sec, 0,  &last_read);
+		pr_info("my_time: call %s ", __func__);
+		sprintf(buf, "Now: %04ld-%02d-%02d %02d-%02d-%02d\n",
+			1900 + last_read.tm_year, 1 + last_read.tm_mon, last_read.tm_mday,
+			last_read.tm_hour, last_read.tm_min, last_read.tm_sec);
+		sprintf(buf + strlen(buf), "Time in sec from previous read:: %lu\n",
+				(jiffies_now - jiffies_stamp) / HZ);
+		time_to_tm(epoch_now.tv_sec - (jiffies_now - jiffies_stamp) / HZ, 0,
+			&last_read);
+		sprintf(buf + strlen(buf),
+			"Last read time (absolute): %04ld-%02d-%02d %02d-%02d-%02d\n",
+			1900 + last_read.tm_year, 1 + last_read.tm_mon, last_read.tm_mday,
+			last_read.tm_hour, last_read.tm_min, last_read.tm_sec);
+		jiffies_stamp = jiffies_now;
+	} else {
+		sprintf(buf, "Overflow detected. Re-init now.");
+		jiffies_stamp = jiffies_now;
+	}
+	return strlen(buf);
+}
+
+static ssize_t my_store(struct class *class, struct class_attribute *attr,
+						const char *buf, size_t count)
+{
+	int ret_value;
+
+	pr_info("my_time: call %s", __func__);
+	ret_value = sscanf(buf, "%lu", &delta_in_ms);
+
+	if (ret_value != 1) {
+		pr_info("my_time: call %s, can't read delay value!\n", __func__);
+	}
+
+	if (delta_in_ms < MIN_DELAY_IN_MS) {
```

Comment:
> if (delta_in_ms < MIN_DELAY_IN_MS) {

In this case the timer will be restarted with default value, correct?
So the previous delay period successfully set before would be dropped.